### PR TITLE
Added mailchimp_automations submodule and updated api docs

### PR DESCRIPTION
**Pull request by a-fro**
Per our email conversation, here's the initial, working implementation of the mailchimp automations module added to the suite. I've tested this with an entityform on our site, and it works great! 

Feedback welcome!

### CODE REVIEW

**Comment by Greg-Boggs** [comment]
The code looks very solid. We'll need to apply coding standards fixes. I've pasted them here. Let me know if you can't get to them and I'm happy to pitch in.

## FILE: .../mailchimp_automations/includes/mailchimp_automations.entity.inc

## FOUND 10 ERRORS AFFECTING 8 LINES

  8 | ERROR | [x] Missing class doc comment
 10 | ERROR | [ ] Class property $mailchimp_automation_entity_id
    |       |     should use lowerCamel naming without underscores
 10 | ERROR | [ ] There must not be more than one property declared
    |       |     per statement
 10 | ERROR | [ ] Visibility must be declared on property
    |       |     "$mailchimp_automation_entity_id"
 13 | ERROR | [ ] Class property $entity_type should use lowerCamel
    |       |     naming without underscores
 15 | ERROR | [ ] Class property $email_property should use
    |       |     lowerCamel naming without underscores
 16 | ERROR | [ ] Class property $list_id should use lowerCamel
    |       |     naming without underscores
 17 | ERROR | [ ] Class property $workflow_id should use lowerCamel
    |       |     naming without underscores
 18 | ERROR | [ ] Class property $workflow_email_id should use
    |       |     lowerCamel naming without underscores
 19 | ERROR | [ ] Class property $entity_path should use lowerCamel

##     |       |     naming without underscores

## PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

## FILE: ...imp_automations/includes/mailchimp_automations.ui_controller.inc

## FOUND 1 ERROR AFFECTING 1 LINE

 16 | ERROR | Public method name
    |       | "MailChimpAutomationUIController::hook_menu" is not in

##     |       | lowerCamel format

## FILE: ...ggs/projects/mailchimp_automations/mailchimp_automations.install

## FOUND 14 ERRORS AFFECTING 14 LINES

  14 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  19 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  22 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  28 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  34 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  41 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  48 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  55 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  62 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  69 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  76 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  84 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
  97 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators
 100 | ERROR | Do not use t() in hook_schema(), this will only
     |       | generate overhead for translators

**Comment by a-fro** [comment]
Thanks Greg! Sorry about that. I thought I'd handled the coding issues. I can get back to it tomorrow morning.

**Comment by a-fro** [comment]
I pushed the hook_schema translate change. I'm not sure how you guys would like me to handle the other errors, as the code is based on mailchimp_activity and they show the exact same list of errors. It seems like that's the pattern you all have used in the mailchimp module suite. Please advise. Thanks. 

One other nagging question I've had is whether it should be called mailchimp_automation or mailchimp_automations. The submodules use both singular and plural. If the former, please let me know and I'll change it accordingly.

**Comment by Greg-Boggs** [comment]
Ideally, we'd make all of MC pass coder so that we can use coder to review contributions more easily. But, we're happy to accept the PR without a perfect coder result. It will need to be improved in the future.

 It's been on our list for a while, but we haven't done it yet made MailChimp pass coder review because there are several other large projects going on including mailchimp_commerce and the V2 API deprecation.

I don't have any preference between the two name choices. So, whichever you've used already works for me. :)

**Comment by a-fro** [comment]
That certainly makes sense. Other than the class doc comment, it's not actually clear to me that the other "errors" related to lowerCamel are correct, from a Drupal perspective. It might be better for the TS team to decide how you want to handle those, since it's related to all the modules in the suite. Thanks for the offer to improve it in the future.

**Comment by Greg-Boggs** [comment]
I had to look this one up. It seems the Drupal perspective is lowerCamel which is a change that came from Drupal 8 best practices. I'll open a separate PR with coding standards fixes after we accept this one.

**Comment by a-fro** [comment]
Thanks @Greg-Boggs. Any word on timeline for this and the multiple key PR? I need to come up with a plan for our stakeholders and communicate about timing. Thanks!

**Comment by ruscoe** [comment]
@a-fro I'm reviewing both now :)

File: mailchimp.api.php [hunk 1]
```diff
@@ -56,3 +56,43 @@ function hook_mailchimp_subscribe_user($list_id, $email, $merge_vars) {
 function hook_mailchimp_unsubscribe_user($list_id, $email) {
 
 }
+
```
Comment (ruscoe) [inline mailchimp.api.php]: Would it make sense to put these in `modules/mailchimp_automations/mailchimp_automations.api.php`?

File: mailchimp.api.php [hunk 2]
```diff
@@ -56,3 +56,43 @@ function hook_mailchimp_subscribe_user($list_id, $email, $merge_vars) {
...
+ *   The MailchimpAutomationEntity object.
+ * @param string $email
+ *   The email_property value from the MailchimpAutomationEntity.
+ * @param object $wrapped_entity
+ *   The EntityMetadataWrapper for the triggering entity.
+ */
+function hook_mailchimp_automations_workflow_email_triggered($automation_entity, $email, $wrapped_entity) {
+
+}
+
```
Comment (ruscoe) [inline mailchimp.api.php:98]: Would it make sense to put these in `modules/mailchimp_automations/mailchimp_automations.api.php`?

File: modules/mailchimp_automations/includes/mailchimp_automations.admin.inc [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            $automation_options = mailchimp_automations_options_list($automations);
+            $form['drupal_entity']['workflow_id'] = array(
+              '#title' => t('Workflow ID'),
+              '#type' => 'select',
+              '#required' => TRUE,
+              '#description' => t('Select the workflow automation that you created in MailChimp. Untitled workflows are not included.'),
+              '#options' => $automation_options,
+              '#default_value' => $entitynotnull ? $mailchimp_automations_entity->workflow_id : 0,
+              '#maxlength' => 127,
+              '#weight' => -5,
```
Comment (ruscoe) [inline modules/mailchimp_automations/includes/mailchimp_automations.admin.inc]: This `#weight` is the same as `$form['drupal_entity']['list_id']['#weight']`, which makes the Workflow ID field appear above the MailChimp List field. It's a really minor thing, but it's weird to have a field appear above another when loaded via AJAX. Can this be changed to `-4` and `$form['drupal_entity']['workflow_email_id']` be changed to `-3`?

**Comment by ruscoe** [comment]
I added comments for a couple of minor things, but overall looks great. It's clear that a lot of work went into this PR; thank you for putting in the effort.

File: mailchimp.api.php (hunk 2 above)
Comment (a-fro) [reply mailchimp.api.php:98]: That was my first instinct, and what I did initially. However, it appears the pattern your team has used to this point it to put api hook documentation for all the mailchimp modules in this document (@see hook_mailchimp_lists_mergevars_alter), so I moved it here. Redhen looks like it does the same. However, I checked media and they do it the other way. Seems like there's not a clear standard. Please advise.

File: modules/mailchimp_automations/includes/mailchimp_automations.admin.inc (hunk 3 above)
Comment (a-fro) [reply modules/mailchimp_automations/includes/mailchimp_automations.admin.inc]: Will do. Thanks!

File: mailchimp.api.php (hunk 2 above)
Comment (ruscoe) [reply mailchimp.api.php:98]: Good point. In that case, I think for consistency these hooks are fine in mailchimp.api.php.

**Comment by a-fro** [comment]
@ruscoe thanks for the code review and feedback! I'm happy to be able to help. 

~~I added a question above about the api file, and also added an additional commit to support multiple api keys on the entity form #ajax since it sounds like that is going to be accepted. Let me know if you see other issues.~~ Commit removed based on implementation change.

**Comment by ruscoe** [comment]
Thanks for all the work on this! It's in the 7.x-4.x branch now. Sorry for the delay in merging it.

**Pull request merged**